zipline/pipeline/loaders/quarter_estimates.py

```diff
@@ -0,0 +1,266 @@
+                SID_FIELD_NAME
+            ) == sid]
+            # Determine where we think quarters are changing for this sid.
+            qtr_shifts = sid_result[
+                sid_result[SHIFTED_NORMALIZED_QTRS] !=
+                sid_result[SHIFTED_NORMALIZED_QTRS].shift(1)
+            ]
+            # Iterate backwards making adjustments. No adjustment for 1st
+            # quarter.
+            for row_indexer in list(reversed(qtr_shifts.index))[:-1]:
+                # We want to write estimates for this row's quarter over
+                # everything that comes before this quarter when we are at
+                # the date when this quarter starts.
+                qtr_start_idx = last.index.get_loc(row_indexer[0])
+                quarter = qtr_shifts.loc[row_indexer][SHIFTED_NORMALIZED_QTRS]
+                adjustments[qtr_start_idx] = \
+                    [overwrite(
+                        0,
+                        qtr_start_idx - 1,  # overwrite thru last qtr
+                        sid_idx,
+                        sid_idx,
+                        last[column_name,
+                             quarter,
+                             sid][:qtr_start_idx].values)]
+
+        return AdjustedArray(
+                col_result.values.astype(column.dtype),
+                mask,
+                dict(adjustments),
+                column.missing_value,
+            )
+
+    def load_adjusted_array(self, columns, dates, assets, mask):
+        # TODO: how can we enforce that datasets have the num_quarters
+        # attribute, given that they're created dynamically?
+        groups = groupby(lambda x: x.dataset.num_quarters, columns)
+        groups_columns = dict(groups)
+        if (pd.Series(groups_columns.keys()) < 0).any():
+            raise ValueError("Must pass a number of quarters >= 0")
+        out = {}
+        self.estimates[NORMALIZED_QUARTERS] = normalize_quarters(
+            self.estimates[FISCAL_YEAR_FIELD_NAME],
+            self.estimates[FISCAL_QUARTER_FIELD_NAME],
+        ).astype(float)
+        for num_quarters, columns in groups_columns.iteritems():
+            # The column's dataset is itself dynamic and the mapping we
+            # actually want is to its dataset's parent's column name.
+            name_map = {c: self.base_column_name_map[
+                            getattr(c.dataset.__base__, c.name)
+                        ] for c in columns}
+            # Determine the last piece of information we know for each column
+            # on each date in the index for each sid and quarter.
+            last_per_qtr = last_in_date_group(
+                self.estimates, True, dates, assets,
+                extra_groupers=[NORMALIZED_QUARTERS]
+            )
+
+            # Forward fill values for each quarter.
+            ffill_across_cols(last_per_qtr, columns)
+            # Stack quarter and sid into the index
+            stacked_last_per_qtr = last_per_qtr.stack(
```

Comment:
I couldn't figure out any other way to perform the filtering/grouping operations in `load_quarters`.